src/Tutor.Web/Controllers/Analytics/AnalyticsController.cs

```diff
@@ -15,25 +21,53 @@ namespace Tutor.Web.Controllers.Analytics
     public class AnalyticsController : ControllerBase
     {
         private readonly ILearnerRepository _learnerRepository;
+        private readonly IDomainRepository _domainRepository;
         private readonly IEventStore _eventStore;
         private readonly IMapper _mapper;
 
-        public AnalyticsController(ILearnerRepository learnerRepository, IEventStore eventStore, IMapper mapper)
+        public AnalyticsController(ILearnerRepository learnerRepository, IDomainRepository domainRepository, IEventStore eventStore, IMapper mapper)
         {
             _learnerRepository = learnerRepository;
+            _domainRepository = domainRepository;
             _eventStore = eventStore;
             _mapper = mapper;
         }
 
         [HttpGet("learner-progress")]
         public ActionResult<PagedResult<LearnerProgressDto>> GetProgress([FromQuery] int page, [FromQuery] int pageSize, [FromQuery] int groupId)
         {
-            var task = _learnerRepository.GetLearnersAsync(page, pageSize, groupId);
+            var task = _learnerRepository.GetLearnersWithMasteriesAsync(page, pageSize, groupId);
             task.Wait();
             var results = task.Result.Results.Select(progress => _mapper.Map<LearnerProgressDto>(progress)).ToList();
             return Ok(new PagedResult<LearnerProgressDto>(results, task.Result.TotalCount));
         }
 
+        [HttpGet("kc-statistics")]
+        public ActionResult<List<KcStatisticsDto>> GetKcStatistics([FromQuery] int unitId, [FromQuery] int groupId)
+        {
+            // Should refactor. Need to research appropriate patterns for this case. AnalyticsService seems to be a simple solution.
+            var unit = _domainRepository.GetUnit(unitId);
+            var learnerIds = _learnerRepository.GetByGroupId(groupId).Select(l => l.Id).ToList();
+
+            var events = _eventStore.GetKcEvents(unit.KnowledgeComponents.Select(kc => kc.Id).ToList(), learnerIds);
+            
+            return Ok(CalculateKcStatistics(unit, events));
+        }
+
+        private static List<KcStatisticsDto> CalculateKcStatistics(KnowledgeUnit unit, List<KnowledgeComponentEvent> events)
+        {
+            return unit.KnowledgeComponents.Select(kc =>
+                new KcStatisticsDto
+                {
+                    KcCode = kc.Code,
+                    KcName = kc.Name,
+                    TotalStarted = events.Where(e => e is SessionLaunched && e.KnowledgeComponentId == kc.Id).Select(e => e.LearnerId).Distinct().Count(),
+                    TotalCompleted = events.Where(e => e is KnowledgeComponentCompleted && e.KnowledgeComponentId == kc.Id).Select(e => e.LearnerId).Distinct().Count(),
+                    TotalPassed = events.Where(e => e is KnowledgeComponentPassed && e.KnowledgeComponentId == kc.Id).Select(e => e.LearnerId).Distinct().Count()
```

Comment:
One additional thing to keep in mind - I would like to determine the AverageTotalTime a KC has been worked on, which we can determine as the difference between the first and last TimeStamp of any events.

Also of interest is the AverageEffectiveTime - which should ideally examine events between SessionStarted and Abandoned, but since Abandoned is unreliable an option is to view the event before the next Started as the AbandonedSubstitute. It's not 100% accurate, but I think it is the best we can do with this data.

Note that both times should be calculated for learners that passed the KC.

Reply:
One thing I'd like to note related to this is that we initially decided to omit time related data like duration from a lot of events because we were unsure what exactly it would mean at the time - but it seems like we have a better idea of what time measurements we'll need now, so it might be a good idea to revisit that. From the CMI5 related events, SessionTerminated, KnowledgeComponentPassed, and KnowledgeComponentSatisfied are supposed to contain duration, but currently don't. Here's a quick idea of some potentially interesting time data we could enable KCM to calculate and add to some events:
- elapsed time since session launch at the time of the event
- total active time for the KC at the time of the event (or total duration of all sessions at the time of the event)
- elapsed time since KC start at the time of the event.

That would make a lot of the statistics mentioned here simple to calculate, without needing to pair up events and find deltas.

Although, here's a more generic issue I see with this - generally, queries that correlate different types of events seem like some of the most interesting ones to make, but I can't imagine how to actually implement such a query (and not just with the Postgres implementation, with any storage in general). In cases like pairing up a SessionTerminated event with the nearest SessionLaunched event to find the duration of the session it seems like it makes more sense to avoid that query by just adding more data into the SessionTerminated events. But something like 'Which AIs most often cause learners to switch back to IIs?' or 'How many encouraging messages were shown to a learner in the same session that the learner completed or passed a KC?' are perfectly logical and useful questions, and we have all the data to answer them - but I genuinely have no idea how to get to the answers from the raw event data.

Also, what I'm gathering from this is that the SessionAbandoned event as it currently is is effectively useless. Would it be worth it to try to improve it a bit, since the whole solution for that with sockets is on pause anyway? Maybe do something like store a 'LastActivity' timestamp in KCM, and use that as the timestamp for SessionAbandoned instead.